addons/crm/models/crm_lead.py

```diff
@@ -930,15 +931,91 @@ def action_schedule_meeting(self):
         action['context'] = {
-            'default_opportunity_id': self.id if self.type == 'opportunity' else False,
+            'default_opportunity_id': current_opportunity_id,
             'default_partner_id': self.partner_id.id,
             'default_partner_ids': partner_ids,
             'default_team_id': self.team_id.id,
             'default_name': self.name,
         }
+
+        # 'Smart' calendar view : get the most relevant time period to display to the user.
+        if current_opportunity_id:
+            smart_mode, smart_initial_date = self._get_opportunity_meeting_view_parameters(current_opportunity_id)
+            action['context'].update({'default_mode': smart_mode, 'initial_date': smart_initial_date})
+
         return action
 
+    def _get_opportunity_meeting_view_parameters(self, opportunity_id):
+        """ Return the most relevant parameters for calendar view when viewing meetings linked to an opportunity.
+            If there are any meetings that are not finished yet, only consider those meetings,
+            since the user would prefer no to see past meetings. Otherwise, consider all meetings.
+            Allday events datetimes are used without taking tz into account.
+            -If there is no event, return week mode and false (The calendar will target 'now' by default)
+            -If there is only one, return week mode and date of the start of the event.
+            -If there are several events entirely on the same week, return week mode and start of first event.
+            -Else, return month mode and the date of the start of first event as initial date. (If they are
+            on the same month, this will display that month and therefore show all of them, which is expected)
+
+            :param opportunity_id: id of the opportunity of which we want to see the meetings.
+            :return tuple mode, initial_date:
+                - mode: selected mode of the calendar view, 'week' or 'month'
+                - initial_date: date of the start of the first relevant meeting. The calendar will target that date.
+        """
+        meeting_results = self.env["calendar.event"].search_read([('opportunity_id', 'in', [opportunity_id])], ['start', 'stop', 'allday'])
+        if not meeting_results:
+            return "week", False
+
+        user_tz = self.env.user.tz or self.env.context.get('tz')
+        user_pytz = pytz.timezone(user_tz) if user_tz else pytz.utc
+
+        # meeting_dt will contain one tuple of datetimes per meeting : (Start, Stop) as per user time zone.
+        meeting_dt = []
+        now_dt = datetime.now().astimezone(user_pytz).replace(tzinfo=None)
+
+        # When creating an allday meeting, whatever the TZ, it will be stored the same e.g. 00.00.00->23.59.59 in utc or
+        # 08.00.00->18.00.00. Therefore we must not put it back in the user tz but take it raw.
+        for m in meeting_results:
```

Comment:
we prefer complete naming
`for meeting in metting_results`

Reply:
Done in different lists, and clarified naming (...s for structures ... for element)